File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/CustomMenuSingleAction.java
package jp.hazuki.yuzubrowser.action.item;

import android.content.Context;
import android.os.Parcel;
import android.os.Parcelable;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.core.JsonToken;

import java.io.IOException;

import jp.hazuki.yuzubrowser.R;
import jp.hazuki.yuzubrowser.action.ActionList;
import jp.hazuki.yuzubrowser.action.SingleAction;
import jp.hazuki.yuzubrowser.action.view.ActionActivity;
import jp.hazuki.yuzubrowser.action.view.ActionListActivity;
import jp.hazuki.yuzubrowser.action.view.ActionListActivity.OnActionListActivityResultListener;
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo;

public class CustomMenuSingleAction extends SingleAction implements Parcelable {
    private static final String FIELD_NAME_ACTIONLIST = "0";
    private ActionList mActionList;

    public CustomMenuSingleAction(int id, JsonParser parser) throws IOException {
        super(id);

        mActionList = new ActionList();
        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return;
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.getCurrentToken() != JsonToken.FIELD_NAME) return;
                if (!FIELD_NAME_ACTIONLIST.equals(parser.getCurrentName())) {
                    parser.skipChildren();
                    continue;
                }
                mActionList.loadAction(parser);
            }
        }
    }

    @Override
    public void writeIdAndData(JsonGenerator generator) throws IOException {
        generator.writeNumber(id);
        generator.writeStartObject();
        mActionList.writeAction(FIELD_NAME_ACTIONLIST, generator);
        generator.writeEndObject();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeParcelable(mActionList, flags);
    }

    protected CustomMenuSingleAction(Parcel source) {
        super(source.readInt());
        mActionList = source.readParcelable(ActionList.class.getClassLoader());
    }

    public static final Creator<CustomMenuSingleAction> CREATOR = new Creator<CustomMenuSingleAction>() {
        @Override
        public CustomMenuSingleAction createFromParcel(Parcel source) {
            return new CustomMenuSingleAction(source);
        }

        @Override
        public CustomMenuSingleAction[] newArray(int size) {
            return new CustomMenuSingleAction[size];
        }
    };

    @Override
    public StartActivityInfo showMainPreference(ActionActivity context) {
        return showSubPreference(context);
    }

    @Override
    public StartActivityInfo showSubPreference(ActionActivity context) {
        return new ActionListActivity.Builder(context)
                .setTitle(R.string.action_select_menu)
                .setActionNameArray(context.getActionNameArray())
                .setDefaultActionList(mActionList)
                .setOnActionListActivityResultListener(new OnActionListActivityResultListener() {
                    @Override
                    public void onActionListActivityResult(Context context, ActionList actionlist) {
                        mActionList = actionlist;
                    }
                })
                .makeStartActivityInfo();
    }

    public ActionList getActionList() {
        return mActionList;
    }
}

File: app/src/main/java/jp/hazuki/yuzubrowser/action/item/CustomMenuSingleAction.kt
/*
 * Copyright (C) 2017 Hazuki
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package jp.hazuki.yuzubrowser.action.item

import android.os.Parcel
import android.os.Parcelable
import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import jp.hazuki.yuzubrowser.R
import jp.hazuki.yuzubrowser.action.ActionList
import jp.hazuki.yuzubrowser.action.SingleAction
import jp.hazuki.yuzubrowser.action.view.ActionActivity
import jp.hazuki.yuzubrowser.action.view.ActionListActivity
import jp.hazuki.yuzubrowser.utils.app.StartActivityInfo
import java.io.IOException

class CustomMenuSingleAction : SingleAction, Parcelable {
    var actionList: ActionList
        private set

    @Throws(IOException::class)
    constructor(id: Int, parser: JsonParser?) : super(id) {

        actionList = ActionList()
        if (parser != null) {
            if (parser.nextToken() != JsonToken.START_OBJECT) return
            while (parser.nextToken() != JsonToken.END_OBJECT) {
                if (parser.currentToken != JsonToken.FIELD_NAME) return
                if (FIELD_NAME_ACTION_LIST != parser.currentName) {
                    parser.skipChildren()
                    continue
                }
                actionList.loadAction(parser)
            }
        }
    }

    @Throws(IOException::class)
    override fun writeIdAndData(generator: JsonGenerator) {
        generator.writeNumber(id)
        generator.writeStartObject()
        actionList.writeAction(FIELD_NAME_ACTION_LIST, generator)
        generator.writeEndObject()
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeParcelable(actionList, flags)
    }

    private constructor(source: Parcel) : super(source.readInt()) {
        actionList = source.readParcelable(ActionList::class.java.classLoader)
    }

    override fun showMainPreference(context: ActionActivity): StartActivityInfo? {
        return showSubPreference(context)
    }

    override fun showSubPreference(context: ActionActivity): StartActivityInfo? {
        return ActionListActivity.Builder(context)
                .setTitle(R.string.action_select_menu)
                .setActionNameArray(context.actionNameArray)
                .setDefaultActionList(actionList)
                .setOnActionListActivityResultListener { actionList -> this.actionList = actionList }
                .makeStartActivityInfo()
    }

    companion object {
        private const val FIELD_NAME_ACTION_LIST = "0"

        @JvmField
        val CREATOR: Parcelable.Creator<CustomMenuSingleAction> = object : Parcelable.Creator<CustomMenuSingleAction> {
            override fun createFromParcel(source: Parcel): CustomMenuSingleAction {
                return CustomMenuSingleAction(source)
            }

            override fun newArray(size: Int): Array<CustomMenuSingleAction?> {
                return arrayOfNulls(size)
            }
        }
    }
}
